Sources/AppcuesKit/Presentation/Traits/Appcues/TooltipWrapperView.swift

```diff
@@ -11,38 +11,394 @@ import UIKit
+        }
+    }
+
+    override func applyCornerRadius(_ cornerRadius: CGFloat?) {
+        maxCornerRadius = cornerRadius ?? 0
+    }
+
+    override func applyBorder(color: UIColor?, width: CGFloat?) {
+        guard let color = color, let width = width else { return }
+
+        borderLayer.strokeColor = color.cgColor
+        // Multiply by 2 because the stroke is drawn on the edge of the path. The outer half of the strike is trimmed by the mask,
+        // so this is more efficient then calculating another path inset from the tooltip shape path specifically for the border.
+        borderLayer.lineWidth = width * 2
+        contentWrapperView.layoutMargins = UIEdgeInsets(top: width, left: width, bottom: width, right: width)
+    }
+
+    private func verticalPosition(for targetRectangle: CGRect) -> CGRect {
+        let distance = distanceFromTarget
+
+        let safeBounds = bounds.inset(by: safeAreaInsets)
+
+        var targetFrame = CGRect(
+            x: 0,
+            y: 0,
+            width: preferredWidth ?? preferredContentSize?.width ?? Self.defaultMaxWidth,
+            height: preferredContentSize?.height ?? 1)
+
+        // Account for border size
+        targetFrame.size.height += contentWrapperView.layoutMargins.top + contentWrapperView.layoutMargins.bottom
+        // Account for safe area space allocated to the pointer
+        targetFrame.size.height += pointerSize?.height ?? 0
+ 
+        // Cap width to not exceed screen width
+        targetFrame.size.width = min(targetFrame.size.width, safeBounds.width - layoutMargins.left - layoutMargins.right)
+
+        // Determine vertical positioning for targetFrame
+        let spaceAbove = targetRectangle.minY - safeBounds.minY - distance
+        let spaceBelow = safeBounds.maxY - targetRectangle.maxY - distance
+        let excessSpaceAbove = spaceAbove - targetFrame.height
+        let excessSpaceBelow = spaceBelow - targetFrame.height
+
+        if preferredPosition == .top && excessSpaceAbove > 0 {
+            // Position tooltip above the target rectangle
+            targetFrame.origin.y = targetRectangle.minY - targetFrame.height - distance
+            actualPosition = .top
+        } else if preferredPosition == .bottom && excessSpaceBelow > 0 {
+            // Position tooltip below the target rectangle
+            targetFrame.origin.y = targetRectangle.maxY + distance
+            actualPosition = .bottom
         } else {
-            // default to a bottom anchored sheet
-            activeConstraints = [
-                // general position
-                contentWrapperView.bottomAnchor.constraint(equalTo: bottomAnchor),
-                contentWrapperView.leadingAnchor.constraint(equalTo: readableContentGuide.leadingAnchor),
-                contentWrapperView.trailingAnchor.constraint(equalTo: readableContentGuide.trailingAnchor),
+            // TODO: Should this consider a switch to horizontal if there's not enough vertical space on either side?
```

Comment:
good question - does the current "shrink height if too tall to fit" mean it would result in scrolling content inside?

Reply:
Yes, it would scroll

Reply:
gotcha. I think it would be pretty cool if it could switch to leading or trailing tip in that case, but also think that seems totally fine to push down the road and see what feedback we can gather on it.